python/src/uagents/context.py

```diff
@@ -400,19 +454,35 @@ async def send_raw(
         env.encode_payload(json_message)
         env.sign(self._identity)
 
-        try:
-            async with aiohttp.ClientSession() as session:
-                async with session.post(
-                    endpoint,
-                    headers={"content-type": "application/json"},
-                    data=env.json(),
-                ) as resp:
-                    success = resp.status == 200
-                if not success:
-                    self._logger.exception(
+        for endpoint in endpoints:
+            try:
+                async with aiohttp.ClientSession() as session:
```

Comment:
This line should ensure that the function will return as soon as the message is delivered successfully: https://github.com/fetchai/uAgents/pull/150/files#diff-3428a4172029b3a9d519f8cfc73765c0a383e8c65a89ce53879a4db2eccbdb7eR467.

So I think we are in agreement that the function should deliver a maximum of one message. I can add a test case that makes sure this is the case. Having multiple duplicate messages in flight at once would seem to add a lot more complexity that I had in mind at this point.

Reply:
After revisiting this, I think code is actually correct as is. :smile: I was confused for a moment because it's not trivial, but the async context management here simply allows other ongoing tasks to run concurrently, but the `session.post()` must be awaited behind the scenes because the `resp.status` property is immediately available upon entering the context. But let me know if I missed something or if you disagree!

Reply:
Thanks, the test proved what I couldn't deduct from the code. The async stack is a beast in its own right 😄